filebeat/channel/runner.go

```diff
@@ -195,3 +242,22 @@ func setOptional(to mapstr.M, key string, value string) {
 		_, _ = to.Put(key, value)
 	}
 }
+
+func validateDataStreamType(dataStreamType string) bool {
+	validTypes := map[string]bool{"logs": true, "metrics": true, "traces": true}
+	_, exists := validTypes[dataStreamType]
+	return exists
+}
+
+// DISCUSS: What should happen to invalid characters
```

Comment:
@felixbarny Any thoughts on this? This is about the validation of the config of an input in filebeat.

Reply:
I guess validating and throwing an error at configuration time makes sense. The downside is that the feedback is not always immediate and it may be surprising why log shipping fails when, for example, changing the config from `dataset: myapp` to `dataset: my-app` (`-` is disallowed in dataset).

An alternative would be to sanitize instead of validate (`my-app` -> `my_app`).

The reroute processor validates statically configured datasets and prevents creating or updating a processor with an invalid value but sanitizes values from field references, as the user creating the reroute processor can't control the values for `{{service.name}}`, for example.

In that light, it would be consistent to validate in the Filebeat config, also.

For reference, this is the sanitization logic in the reroute processor:
https://github.com/elastic/elasticsearch/blob/5383682d82d285dc55da2d7035fbdbca47794439/modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RerouteProcessor.java#L212-L219

This is also used for validation by checking if the sanitized value is equal to the original one.

Here are the patterns for the allowed characters: https://github.com/elastic/elasticsearch/blob/main/modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RerouteProcessor.java#L195-L196

Reply:
The IndexProcessor (https://github.com/elastic/beats/pull/35287/files#diff-66daa9cc14bccc64d8c8039e1409f67ceeb93bb14bbe93eba11270a42b8fadbfR185) is very generic. Sanitisation there would be tricky (not impossible) as logic would have to be built in for the specific `data_stream.*` fields.

As this PR only touches the constant config, it means we should do the same validation and not send data if the config is invalid. This should not be suprising. It could hit a user by surprise assuming dynamic config reloading is used and then an invalid dataset value is used. But already today, if a config during reload is valid, data is not collected so this should be ok.

@cmacknz Ok with this approach that we don't start the input if an invalid `data_stream.*` value is set?

Reply:
It is definitely better to not start the input with an invalid data stream, if we just log it then it won't be as obvious to the user.